sfs/time/drivingfunction.py

```diff
@@ -251,7 +252,191 @@ def apply_delays(signal, delays):
+    max_order : int
+        Ambisonics order
+    c : float, optional
+        Speed of sound
+
+    Returns
+    -------
+    delay : float
+        Overall delay (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : dictionary
+        Second-order section filters 
+    phaseshift : float
+        Phase shift
+
+    """
+    if max_order is None:
+        max_order =_max_order_circular_harmonics(len(x0), max_order)
+    if c is None:
+        c = defs.c
+
+    x0 = util.asarray_of_rows(x0)
+    xs = util.asarray_1d(xs)
+    phi0, _, _ = util.cart2sph(*x0.T)
+    phi, _, r = util.cart2sph(*xs)
+
+    delay = (r-r0)/c
+    weight = 1/2/np.pi/r
+    sos = {}
+    for m in range(0, max_order+1):
+        _, p, k = sig.besselap(m, norm='delay')
+        # TODO: modify "sig.besselap" for very high orders (>150)
+        s0 = (c/r)*p
+        sinf = (c/r0)*p
+        z0 = np.exp(s0/fs)
+        zinf = np.exp(sinf/fs)
+        # TODO: select matched-z or bilinear transform
+        if normalize:
+            k = _normalize_gain(s0, sinf, z0, zinf, fs=fs)
+        else:
+            k = 1
+        sos[m] = sig.zpk2sos(z0,zinf,k,pairing='nearest')
+        # TODO: normalize the SOS filters individually?
+    phaseshift = phi0 - phi
+    return delay, weight, sos, phaseshift
+
+
+def nfchoa_driving_signals(delay, weight, sos, phaseshift, signal, max_order=None, fs=44100):
+    """Get NFC-HOA driving signals per secondary source.
+
+    Parameters
+    ----------
+    delay : float
+        Overall delay (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : dictionary
+        Second-order section filters 
+    phaseshift : (C,) array_like
+        Phase shift
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    fs: int, optional
+        Sampling frequency in Hertz.
+
+    Returns
+    -------
+    driving_signals : (N, C) numpy.ndarray
+        Driving signals.
+    t_offset : float
+        Simulation point in time offset (seconds).
+
+    """
+    if max_order is None:
+        max_order =_max_order_circular_harmonics(len(phaseshift), max_order)
+
+    delay = util.asarray_1d(delay)
+    weight = util.asarray_1d(weight)
+#    TODO : check FOS/SOS structure
+
+    N = len(phaseshift)
+    L = len(signal)
+    d = np.zeros((L,N),dtype='complex128')
+
+    modal_response = sig.sosfilt(sos[0], signal)
+    for l in range(N):
+        d[:,l] += modal_response
+    for m in range(1, max_order+1):
+        modal_response = sig.sosfilt(sos[np.abs(m)], signal)
+        for l in range(N):
+            d[:,l] += modal_response * 2 * np.cos(m*phaseshift[l])
+    t_offset = delay[0]
+    return np.real(d) * weight, fs, t_offset
+    
+def _max_order_circular_harmonics(N, max_order):
+    """Compute order of 2D HOA."""
+    return (N-1) // 2 if max_order is None else max_order
```

Comment:
This is one of the functions we need for the time domain and the mono-frequent implementation. It is already implemented  [mono/drivingfunction.py](https://github.com/sfstoolbox/sfs-python/blob/master/sfs/mono/drivingfunction.py#L678), but in a slight different way:
```python
def _max_order_circular_harmonics(N, max_order):
    """Compute order of 2D HOA."""
    return N // 2 if max_order is None else max_order
```
Having a look at Ahrens (2012), p. 132 there should also be a distinction of cases for even and odd numbers:
```
     / N/2 - 1,   even nls
M = <
     \ (N-1)/2    odd nls
```

Another problem is that this only holds for 2D and 2.5D. For 3D we would have:
```
      _____
M = \|nls/2
```
see also [nfchoa_order.m](https://github.com/sfstoolbox/sfs-matlab/blob/master/SFS_general/nfchoa_order.m).
As we don't have 3D at the moment we could stick with the function name and change it later if we need 3D.

TODOS:
* find a common place and define only one `_max_order_circular_harmonics(N, max_order)` function
* add distinction between odd and even N
* add a reference to the help message

Reply:
Hi @narahahn,
I just realized that `(N-1) // 2` is already the correct implementation of the max order function.
And the 2D vs. 3D problem can be solved by naming the 3D case `_max_order_spherical_harmonics`.
From the mentioned TODOs, then only these remain:
* find a common place and define only one `_max_order_circular_harmonics(N, max_order)` function
* add a reference to the help message

Reply:
I agree. Maybe in `sfs.util`?

Reply:
Yes, I think `sfs.util` would be currently the best fit.

Reply:
While you are at it, you could replace the `if`/`else` expression with an `if`/`else` statement, I think this would be easier to understand.

Then it should also become more obvious that you should remove the separate check for `None` before calling the function.

Reply:
I tried to implement this with ef78d76, where I moved `_max_order_circular_harmonics` to `util`,  renamed it to `max_order_circular_harmonics` and added documentation.

For the documentation I would need some help. At the moment I added `page 132 of [Ahrens2012]_`, but it would be nicer to replace this by the actual equation. As I don't have a copy of the book here I cannot do this myself.

Reply:
Do you need the equation number? It's (4.26)